File: src/main/java/lk/tech/tgcontrollerclient/utils/KeyManager.java
package lk.tech.tgcontrollerclient.utils;

import lombok.NoArgsConstructor;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.UUID;

@NoArgsConstructor
public enum KeyManager {

    INSTANCE;

    private String key;
    private Path KEY_FILE;

    public void init() {
        try {
            Path CONFIG_DIR = Paths.get(System.getenv("APPDATA"), "Desktop Control Telegram");
            KEY_FILE = CONFIG_DIR.resolve("client.key");

            Files.createDirectories(CONFIG_DIR);
            key = loadOrGenerateKey();

        } catch (Exception e) {
            throw new RuntimeException("KeyManager init failed", e);
        }
    }

    public String key() {
        return key;
    }

    public String regenerateKey() {
        try {
            String newKey = generateKey();
            Files.writeString(KEY_FILE, newKey, StandardCharsets.UTF_8);
            key = newKey;
            return newKey;
        } catch (IOException e) {
            throw new RuntimeException("Unable to regenerate key", e);
        }
    }

    private String loadOrGenerateKey() throws IOException {
        if (Files.exists(KEY_FILE)) {
            String existing = Files.readString(KEY_FILE, StandardCharsets.UTF_8).trim();
            if (!existing.isEmpty()) return existing;
        }
        return regenerateKey();
    }

    private String generateKey() {
        return UUID.randomUUID().toString();
    }
}

File: utils/key_manager.go
package utils

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/google/uuid"
)

type KeyManager struct {
	key     string
	keyFile string
}

var KM = &KeyManager{} // аналог ENUM INSTANCE

// Init — инициализация KeyManager
func (km *KeyManager) Init() error {
	appData := os.Getenv("APPDATA")
	if appData == "" {
		return fmt.Errorf("APPDATA env var is empty")
	}

	configDir := filepath.Join(appData, "Desktop Control Telegram")
	km.keyFile = filepath.Join(configDir, "client.key")

	if err := os.MkdirAll(configDir, 0755); err != nil {
		return fmt.Errorf("cannot create config dir: %w", err)
	}

	key, err := km.loadOrGenerateKey()
	if err != nil {
		return fmt.Errorf("loadOrGenerateKey failed: %w", err)
	}

	km.key = key
	return nil
}

// Key — геттер
func (km *KeyManager) Key() string {
	return km.key
}

// RegenerateKey — пересоздаёт ключ и сохраняет на диск
func (km *KeyManager) RegenerateKey() (string, error) {
	newKey := km.generateKey()
	err := os.WriteFile(km.keyFile, []byte(newKey), 0644)
	if err != nil {
		return "", fmt.Errorf("unable to regenerate key: %w", err)
	}
	km.key = newKey
	return newKey, nil
}

func (km *KeyManager) loadOrGenerateKey() (string, error) {
	data, err := os.ReadFile(km.keyFile)
	if err == nil {
		existing := string(data)
		if len(existing) > 0 {
			return existing, nil
		}
	}
	return km.RegenerateKey()
}

func (km *KeyManager) generateKey() string {
	return uuid.New().String()
}
